### PR TITLE
Make title editable

### PR DESCRIPTION
**Pull request by varung97**
Issue with this implementation: Title doesn't show `inactive` anymore

### CODE REVIEW

**Comment by varung97** [comment]
Fixed, but if the title isn't changed at all, then it warns that title is a duplicate. Any ideas on how to fix?

File: src/ui/add_card/addActions.js [hunk 1]
```diff
@@ -6,14 +6,18 @@ import ToastStore from "../toast/ToastStore";
...
 };
 
 function saveAddList(title, tagString) {
 	title = title.trim();
 	const tagArr = tagStringToArray(tagString);
 
-	if (isInvalidTitleOrTags(title, tagArr)) {
+	if (isInvalidTitle(title) || isInvalidTags(tagArr)) {
 		ToastStore.isInvalidTitleOrTags = true;
 	}
```
Comment (kabirgh) [inline src/ui/add_card/addActions.js]: Much cleaner than isInvalidTitleOrTags, thanks

File: src/ui/spoiler_card/SpoilerCard.js [hunk 2]
```diff
@@ -6,20 +6,20 @@ import OptionsMenuContainer from "../options_menu/OptionsMenuContainer";
...
 		style={{position: "relative",
 				backgroundColor: "#F5F8FA",
 				color: props.isActive ? "#000000" : "#5C7080",
 				borderLeft: props.isActive ? "8px solid #29A634" : "8px solid #738694",
 				marginBottom: props.marginBottom}}>
 		{/* TODO: make editable */}
 		<div>
-			{props.title + (props.isActive ? "" : " (Inactive)")}
+			<EditableText value={props.title + ((props.editing || props.isActive) ? "" : " (Inactive)")} onChange={props.onUpdateTitle} onConfirm={props.onSaveTitle} onEdit={props.onEditTitle} onCancel={props.onCancelTitleEdit} />
 		</div>
```
Comment (kabirgh) [inline src/ui/spoiler_card/SpoilerCard.js]: You could show each prop on its own line so its easy to read. 
eg. 
```
<Text
    value={...}
    onChange={...}
    onConfirm={...}
/>
```

File: src/ui/spoiler_card/SpoilerCard.js [hunk 3]
```diff
@@ -31,28 +31,31 @@ const SpoilerCard = observer(props =>
...
 	</div>
 );
 
 SpoilerCard.propTypes = {
 	onExpandCollapse: PropTypes.func.isRequired,
 	isExpanded: PropTypes.bool.isRequired,
-	title: PropTypes.string.isRequired,
 	isActive: PropTypes.bool.isRequired,
+	editing: PropTypes.bool.isRequired,
+	title: PropTypes.string.isRequired,
```
Comment (kabirgh) [inline src/ui/spoiler_card/SpoilerCard.js]: bools should be prefixed with is/should or similar. editing  =>  isEditing

File: src/ui/spoiler_card/SpoilerCardContainer.js [hunk 4]
```diff
@@ -10,30 +10,61 @@ import SpoilerCard from "./SpoilerCard";
 @observer
```
Comment (kabirgh) [inline src/ui/spoiler_card/SpoilerCardContainer.js]: Remove TODO

File: src/ui/spoiler_card/SpoilerCardContainer.js [hunk 5]
```diff
@@ -1,7 +1,7 @@
 import React from "react";
 const PropTypes = React.PropTypes;
 import {observer} from "mobx-react";
-import {observable} from "mobx";
+import {observable, computed} from "mobx";
 import spoilerCardActions from "./spoilerCardActions";
```
Comment (kabirgh) [inline src/ui/spoiler_card/SpoilerCardContainer.js]: No need to import computed

File: src/ui/spoiler_card/SpoilerCardContainer.js [hunk 6]
```diff
@@ -10,30 +10,61 @@ import SpoilerCard from "./SpoilerCard";
 @observer
 class SpoilerCardContainer extends React.Component {
 	@observable isExpanded = false;
+	@observable title = null;
 	/* TODO: does 'tags' have to be observable? */
 	@observable tags = null;
+	@observable editing = false;
```
Comment (kabirgh) [inline src/ui/spoiler_card/SpoilerCardContainer.js]: editing  =>  isEditing

File: src/ui/spoiler_card/spoilerCardActions.js [hunk 7]
```diff
@@ -1,23 +1,48 @@
...
+		ToastStore.isInvalidTitleOrTags = true;
+		return false;
+	}
+	else if (addActions.isDuplicateTitle(title)) {
+		ToastStore.isDuplicateTitle = true;
+		return false;
+	} else {
+		return true;
+	}
+}
```
Comment (kabirgh) [inline src/ui/spoiler_card/spoilerCardActions.js]: I'd like not to introduce a dependency on addActions. Do you think shifting isInvalidTitle, isDuplicateTitle to a commonActions file makes more sense?

**Review by kabirgh — COMMENTED** [review]
To work around the new = old marked as duplicate problem, perhaps you could use the index argument supplied to ignore the list being edited.

**Comment by kabirgh** [comment]
Btw let's squash and merge all pull requests. Master branch stays cleaner and individual commits can be found in the pull request that was merged.

File: src/ui/spoiler_card/spoilerCardActions.js (hunk 7 above)
Comment (varung97) [reply src/ui/spoiler_card/spoilerCardActions.js]: Yeah I think that would be good. So maybe a commons folder inside ui?

**Comment by varung97** [comment]
Yeah sure will start doing that from now on
You may also be able to make that the only [way](https://github.com/blog/2141-squash-your-commits) of doing things.

**Comment by varung97** [comment]
Made the requested changes

File: src/ui/common/commonActions.js [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+import MainStore from "../MainStore";
+import ToastStore from "../toast/ToastStore";
+
+module.exports = {
+  tagStringToArray: tagStringToArray,
+	isDuplicateTitle: isDuplicateTitle,
+  isDuplicateTitleSkipIndex: isDuplicateTitleSkipIndex,
+	isInvalidTitle: isInvalidTitle,
+	isInvalidTags: isInvalidTags
+};
```
Comment (kabirgh) [inline src/ui/common/commonActions.js]: Indentation + wrap the object values in 'action'. eg. `isInvalidTags: action(isInvalidTags)` - doing this should provide [better performance as well as better debugging](https://medium.com/@mweststrate/mobx-2-2-explicit-actions-controlled-mutations-and-improved-dx-45cdc73c7c8d#.7a0rdfqga) with MobX devtools.

File: src/ui/common/commonActions.js [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+  return isDuplicateTitleSkipIndex(title, -1);
+}
+
+function isDuplicateTitleSkipIndex(title, index) {
+	const lowerCaseTitle = title.toLowerCase();
+
+	for (let i=0; i<MainStore.lowerCaseTitles.length; i++) {
+		if (i != index && lowerCaseTitle === MainStore.lowerCaseTitles[i]) {
+      indicateDuplicateTitle()
+			return true;
```
Comment (kabirgh) [inline src/ui/common/commonActions.js]: Missing semicolon

**Review by kabirgh — COMMENTED** [review]
Looks good! Since we now have a common folder, seems like OptionStore and ToastStore should be moved there too (which I'm on the fence about since it introduces nesting). What's your opinion on leaving commonActions in the root folder, with panel.js, etc. vs moving the stores into common?

**Comment by varung97** [comment]
Hmm but Toast is a component also right? In which case it should be placed along with the other components, I think.
OptionStore, I think maybe does make sense to move it in... not sure though

File: src/ui/common/commonActions.js (hunk 8 above)
Comment (varung97) [reply src/ui/common/commonActions.js]: Wow Atom mixed tabs and spaces somehow...

**Comment by kabirgh** [comment]
Mistyped `ToastStore` there, meant `MainStore`. Hmmm... Let's move both `MainStore` and `OptionStore` into `common/` too. The structure will then mirror `logic/` as well.

**Comment by varung97** [comment]
Sure I don't mind that.
I'll do it in a bit. Gotta fix these merge conflicts also 😓

**Pull request merged**